cpp/src/arrow/compute/function.h

```diff
@@ -225,6 +233,15 @@ class ARROW_EXPORT Function {
   /// required by the kernel.
   virtual Result<const Kernel*> DispatchBest(std::vector<TypeHolder>* values) const;
 
+  /// \brief Get a function executor with a best-matching kernel
+  virtual Result<std::shared_ptr<FunctionExecutor>> BestExecutor(
+      const std::vector<Datum>& args, const FunctionOptions* options,
+      ExecContext* ctx) const;
```

Comment:
Without looking at the impl, only the description, the signatures here look weird: I'd expect them to take types, not datums, since that's what kernel resolution uses.

Reply:
I can see having these as additional overloads for convenience, but from the Jira, it sounds like it'd be most useful to be able to create an executor from just types?

Reply:
Yes, I would expect a `const std::vector<TypeHolder>&` as well.

Reply:
Also, why does a `ExecContext` need to be passed? Does it go into kernel resolution?

Reply:
Makes sense and I'll look into this. So far I just refractored in a simple way.